source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -3836,6 +3840,32 @@ class ActionButton {
     }
 }
 
+class ActionButtonWithTelemetry extends ActionButton {
+	
+	render(alignment: Enums.ActionAlignment) {
+
+		var useGUID = GUIDHelper.getOrCreate().isGUIDtracked();
+		var guid = GUIDHelper.getOrCreate().getGUID().toString();
+
+        this.action.render();
+        this.action.renderedElement.style.flex = alignment === Enums.ActionAlignment.Stretch ? "0 1 100%" : "0 1 auto";
+        this.action.renderedElement.onclick = (e) => {
+            e.preventDefault();
+			e.cancelBubble = true;
+			
+			if (this.action instanceof SubmitAction && useGUID) {
+				ACLogger.getOrCreate().logEvent("SubmitAction", "Action.Submit", guid);
```

Comment:
- To address your first point, Action.Submit is only logged because that was the only action that was in scope for this iteration of this feature
- Regarding your fourth point, I agree that this would be a better alternative. I see that class "Action" has an "onExecute" property and I'd like to explore adding a hook here, although I am not seeing where this function is instantiated. Could you point me in the right direction for this?
- To address your fifth point, the GUID is used to correlate events on a unique card after that card is rendered. So if I render card 1 then render card 2, each would have a unique GUID. But later if I click a submit button belonging to card 1, then the hook will send a log containing the same GUID as when card 1 was rendered. I can do this at render time because I attach a hook to the onClick listener that holds even after the card is finished rendering.
- Addressing your final point, useGUID is true only if the card currently being rendered has an Input.Rating element on it. Using GUIDHelper to track this condition between the JSON parsing and the button render was used in order to avoid more encroachment into the existing code and avoid adding any global variables. There is a possibility that I could add more of this tracking logic within ACLogger to lessen the impact within existing code, and I plan on exploring this option.

Please let me know your thoughts. Thanks again for your feedback!

Reply:
- Got it, that said I think it's a bit of a missed opportunity to address things in a generic way from the get go
- The code should probably live in the execute() method of the base Action class
- I am not sure I get it. I'll ping you offline to discuss
- Not sure this is the right approach. You seem to be using the term "useGUID" to mean something very specific, so maybe at least give it a more specific name?